test/docker-e2e/e2e_upgrade_test.go

```diff
@@ -0,0 +1,306 @@
+package docker_e2e
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+
+	"celestiaorg/celestia-app/test/docker-e2e/dockerchain"
+
+	"github.com/celestiaorg/tastora/framework/testutil/wait"
+	tastoratypes "github.com/celestiaorg/tastora/framework/types"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+)
+
+// TallyResponse represents the JSON response from the signal tally query
+// We had to define it again because the JSON response type is not a number but string
+type TallyResponse struct {
+	VotingPower      string `json:"voting_power"`
+	ThresholdPower   string `json:"threshold_power"`
+	TotalVotingPower string `json:"total_voting_power"`
+}
+
+// TestCelestiaAppMinorUpgrade tests a simple upgrade from one minor version to another by swapping the binary.
+func (s *CelestiaTestSuite) TestCelestiaAppMinorUpgrade() {
+	if testing.Short() {
+		s.T().Skip("skipping celestia-app minor upgrade test in short mode")
+	}
+
+	tt := []struct {
+		Name                string
+		BaseBinaryVersion   string
+		TargetBinaryVersion string
+	}{
+		{
+			Name:                "v4.0.2-rc2 to v4.0.10",
+			BaseBinaryVersion:   "v4.0.2-rc2",
+			TargetBinaryVersion: "v4.0.10",
+		},
+		{
+			Name:                "v4.0.9-mocha to v4.0.10-mocha",
+			BaseBinaryVersion:   "v4.0.9-mocha",
+			TargetBinaryVersion: "v4.0.10-mocha",
+		},
+		{
+			Name:                "v4.0.9-arabica to v4.0.10-arabica",
+			BaseBinaryVersion:   "v4.0.9-arabica",
+			TargetBinaryVersion: "v4.0.10-arabica",
+		},
+	}
+
+	for _, tc := range tt {
+		s.Run(tc.Name, func() {
+			s.RunMinorUpgradeTest(tc.BaseBinaryVersion, tc.TargetBinaryVersion)
+		})
+	}
+}
+
+// TestCelestiaAppMajorUpgrade tests a major upgrade using the signaling mechanism.
+func (s *CelestiaTestSuite) TestCelestiaAppMajorUpgrade() {
+	if testing.Short() {
+		s.T().Skip("skipping celestia-app major upgrade test in short mode")
+	}
+
+	tt := []struct {
+		Name             string
+		BinaryVersion    string
+		TargetAppVersion uint64
+	}{
+		{
+			Name:             "v2 to v3",
+			BinaryVersion:    dockerchain.GetCelestiaTag(),
+			TargetAppVersion: 3,
+		},
+		{
+			Name:             "v3 to v4",
+			BinaryVersion:    dockerchain.GetCelestiaTag(),
+			TargetAppVersion: 4,
+		},
```

Comment:
if we make a PR to v4.x, its using that PR's branch correct? not an already tagged version that we know works already?

Reply:
good consideration, right now tests are being added to main, so if you create a PR against that branch this test will not run (assume it is merged on main).

However it is possible to run this test from main using images that are built on v4.x 

OR you maintain the test of tests across multiple branches. Personally, I find it more manageable to have one set of tests on main, and just build images on feature branches and run workflows.

But both or reasonable solutions and we can do either 👍

Reply:
The tag is set in the ci here: https://github.com/celestiaorg/celestia-app/blob/50da269d3a5d49190716f2e2f378ba0d4fe677d4/.github/workflows/test.yml#L159

and here when the tag is not set, it falls back to a default image: https://github.com/celestiaorg/celestia-app/blob/50da269d3a5d49190716f2e2f378ba0d4fe677d4/test/docker-e2e/dockerchain/config.go#L90

I agree the test should fail if the tag is not set.
Also it should run on every PR, I think running this test is cheap so we can do that. let's sync on this.